S31: Model and CRUD for Projects

qin-guan opened the pull request:
Created models for Projects and Languages and the following endpoints

```http
GET /Projects
GET /Projects/{projectId}
GET /Projects/{projectId}/Languages
POST /Projects
PUT /Projects/{projectId}
DELETE /Projects/{projectId}
```

qin-guan committed:
Added create, update, delete project endpoints
- Added project languages endpoint
- Added models
- Added Serilog
- Added Refit and IGitHubApi
- Added Permission handlers

julwrites commented on Attributes/AddUserIdAttribute.cs:
```diff
@@ -0,0 +1,31 @@
+            var userId = context.HttpContext.User.FindFirst(ClaimTypes.NameIdentifier)?.Value;
+            if (string.IsNullOrWhiteSpace(userId))
+            {
+                throw new UnauthorizedAccessException();
```
Can this be a different exception? Something like `InvalidParameterException`?

qin-guan replied:
It isn't really an invalid parameter, because the `NameIdentifier` is from the Bearer token in the header.

qin-guan replied:
Where is `InvalidParameterException` from?

julwrites replied:
We can define a custom exception, no? I'm just thinking that `UnauthorizedAccessException` is too vague to be very useful.

qin-guan replied:
Why would it be vague?

julwrites commented on Attributes/AddUserIdAttribute.cs:
```diff
@@ -0,0 +1,31 @@
+            var validId = Guid.TryParse(userId, out var userGuid);
+            if (!validId)
+            {
+                throw new UnauthorizedAccessException();
```
Similarly here, `InvalidIdentifierException`?

qin-guan replied:
Where is `InvalidIdentifierException` from?

julwrites commented on Attributes/GithubRepositoryAttribute.cs:
```diff
@@ -0,0 +1,25 @@
+    {
+        protected override ValidationResult IsValid(object value, ValidationContext validationContext)
+        {
+            if (value is null || string.IsNullOrWhiteSpace(value.ToString())) return ValidationResult.Success;
```
Why are we returning success if the value is null or empty?

qin-guan replied:
It is an optional attribute.

qin-guan replied:
Although I should probably remove `string.IsNullOrWhiteSpace(...)` hmm

julwrites replied:
I don't think it makes a difference whether it's optional or not? Unless a null or whitespace value is considered valid in our context, this seems like it will be quite confusing.

qin-guan replied:
It's to be consistent with the other DataAnnotation attributes, if the parameter is required, it should be indicated with `[Required]`.

julwrites commented on Attributes/GithubRepositoryAttribute.cs:
```diff
@@ -0,0 +1,25 @@
+            if (!Uri.TryCreate(value.ToString(), UriKind.Absolute, out var uri))
+                return new ValidationResult("Invalid Url");
+
+            if (uri.Host != "github.com") return new ValidationResult("The repository can only be hosted on GitHub");
```
The formatting of these if statements is quite inconsistent. Does the linter not catch these?

qin-guan replied:
I think it's because line 13 exceeds 120 chars but line 16 doesn't.

julwrites commented on Controllers/ProjectsController.cs:
```diff
@@ -0,0 +1,287 @@
+        [HttpGet]
+        public async Task<ActionResult<List<ProjectResponseDto>>> ReadProjects()
+        {
+            Debug.Assert(HttpContext.Items["UserId"] != null, "HttpContext.Items['UserId'] != null");
```
We should have error checking outside of debug as well.

qin-guan replied:
The assert is only to prevent ReSharper from showing warnings.

julwrites replied:
Yes, but we should still error check, no?

qin-guan replied:
Check for what?

julwrites commented on Controllers/ProjectsController.cs:
```diff
@@ -0,0 +1,287 @@
+            catch (Exception exception)
+            {
+                Log.Fatal(exception, "Fatal error while creating project");
+                return Problem();
```
What is Problem? Any description of the problem or specifics for the parent to catch and handle?

qin-guan replied:
[Problem](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.controllerbase.problem?view=aspnetcore-5.0). I don't think there's a need to reveal to the client the details of the failure.

julwrites replied:
Not unless we want the client to react to it. Do we?

qin-guan replied:
If it's an unexpected failure, the client just displaying unexpected failure is fine right?

julwrites commented on Controllers/ProjectsController.cs:
```diff
@@ -0,0 +1,287 @@
+            {
+                var project = await _dbContext.Projects.Include(p => p.Languages)
+                    .SingleOrDefaultAsync(p => p.Id == projectId);
+                if (project == default(Project)) return NotFound();
```
For a `PUT` action wouldn't you want to insert it if it's not found?

qin-guan replied:
No, `PUT` should be used to update existing projects only.

julwrites commented on Controllers/ProjectsController.cs:
```diff
@@ -0,0 +1,287 @@
+                    }
+                    catch (ApiException exception)
+                    {
+                        Log.Warning(exception, "User submitted invalid repository link");
```
Why is it invalid if languages are not found?

qin-guan replied:
Generally, if the repo exists, then the API will return languages. There are a few exceptions, like the [Kotlin](https://github.com/Jetbrains/Kotlin) that doesn't seem to return any languages, which I'm not sure why either. Also, helps to reduce queries to the API if I dont query if the repository exists first.

julwrites replied:
Okay, which also means that we won't consider a blank repository as valid for our case?

qin-guan replied:
Hmm, good point 🤔, it will still be valid just that there won't be any languages. I'm not sure how we want to keep our language data in sync with GitHub's.

qin-guan replied:
Opened S37 to address this.

julwrites commented on Extensions/PermissionsExtension.cs:
```diff
@@ -0,0 +1,37 @@
+        {
+            foreach (Permission permission in Enum.GetValues(typeof(Permission)))
+            {
+                await roleManager.AddClaimAsync(role, new Claim("Permission", permission.ToString()));
```
We should have these strings as a const var somewhere.

qin-guan replied:
Yeah, I can add it for the claim type.

julwrites commented on Extensions/RoleManagerExtension.cs:
```diff
@@ -0,0 +1,38 @@
+        {
+            await roleManager.CreateAsync(new ApplicationRole
+            {
+                Name = $"Projects/{project.Id.ToString()}/Owner",
```
Similarly should have these strings as a const var somewhere.

qin-guan replied:
How? I don't think it's necessary for this.

julwrites replied:
Probably want it split up and have `Projects` and `Owner` and `Collaborator` be const var somewhere.

julwrites commented on GitHub/IGitHubApi.cs:
```diff
@@ -0,0 +1,12 @@
+    public interface IGitHubApi
+    {
+        [Get("/repos/{owner}/{repo}/languages")]
+        Task<Dictionary<string, double>> GetRepoLanguagesAsync(string owner, string repo);
```
I'm surprised that `GetRepo` isn't also in this API. Is it not an API we need?

qin-guan replied:
Check my comment for `ProjectsController.cs` line 213.

julwrites replied:
Sure, but what if we do want to query repo existence without languages? Is that something we might want to support?

qin-guan replied:
Opened S37 to address this.

julwrites commented on Models/Permission.cs:
```diff
@@ -0,0 +1,15 @@
+        UpdateTask,
+        DeleteTask,
+        
+        AddUser
```
RemoveUser?

qin-guan replied:
Ah yes.

julwrites commented on Requirements/HasPermissionRequirement.cs:
```diff
@@ -0,0 +1,16 @@
+
+        public HasPermissionRequirement(Permission permission)
+        {
+            Permission = permission;
```
Is this meant to be a predicate or a settor?

qin-guan replied:
What do you mean?

julwrites replied:
Is this function supposed to return a boolean or just set the permission? If it's supposed to set, i think we should rename it to `SetPermissionRequirement`.

julwrites commented on Requirements/HasPermissionRequirementHandler.cs:
```diff
@@ -0,0 +1,47 @@
+                roles.Contains(rc.RoleId) && rc.ClaimType == "Permission" &&
+                rc.ClaimValue == requirement.Permission.ToString()).SingleOrDefaultAsync();
+
+            if (hasClaim != default(IdentityRoleClaim<Guid>)) context.Succeed(requirement);
```
Are we guaranteed that if `context.Succeed` is not called, then it will be a failure?

qin-guan replied:
Yes.

qin-guan committed:
Added Dtos for Tags
- Added CustomClaimTypes
- Updated GithubRepositoryAttribute
- Added RemoveUser permission

qin-guan committed:
Added RoleType
- Renamed HasPermissionRequirement to PermissionRequirement
- Renamed HasPermissionRequirementHandler to PermissionRequirementHandler
